File: code/cpp/q101_symmetric_tree.cpp
/*
 *
 *
 * @author: Ricky Chang
 */

#include <climits>
#include <iostream>
#include <unordered_map>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isFlipped(TreeNode* left, TreeNode* right) {
        if ((left == NULL) && (right == NULL)) {
            return true;
        } else if ((left == NULL) || (right == NULL)) {
            return false;
        }

        if ((left->val == right->val) &&
            isFlipped(left->left, right->right) &&
            isFlipped(left->right, right->left)) {
            return true;
        } else {
            return false;
        }

    }

    bool isSymmetric(TreeNode* root) {
        if (root == NULL) {
            return true;
        }

        return isFlipped(root->left, root->right);
    }
};

File: algo_problemset/python/q101_symmetric_tree.py
"""

101. Symmetric Tree
Easy
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    @staticmethod
    def extractValue(root: TreeNode):
        if root is None:
            return None
        else:
            return root.val

    def isSymmetric(self, root: TreeNode) -> bool:

        if root is None:
            return True

        left_queue = [root.left]
        right_queue = [root.right]

        while (len(left_queue) > 0) and (len(right_queue) > 0):

            left_node = left_queue.pop(0)
            right_node = right_queue.pop(0)
            if self.extractValue(left_node) != self.extractValue(right_node):
                return False

            if left_node is not None:
                left_queue.extend([left_node.left, left_node.right])
            if right_node is not None:
                right_queue.extend([right_node.right, right_node.left])

        if (len(left_queue) > 0) or (len(right_queue) > 0):
            return False
        else:
            return True
